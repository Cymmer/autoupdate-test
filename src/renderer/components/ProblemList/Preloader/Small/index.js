import React from 'react';
import styles from './styles.module.scss';

import { Card, Shine } from '../../../elements';

const PreloaderSmall = () => (
  <Card className={styles.PreloaderSmall}>
    <div className={styles.PreloaderSmall_shineHolder}>
      <Shine className={styles.PreloaderSmall_shine} />
    </div>
  </Card>
);

export default PreloaderSmall;
