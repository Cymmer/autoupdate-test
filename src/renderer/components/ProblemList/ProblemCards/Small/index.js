import React from 'react';
import PropTypes from 'prop-types';
import { getUsedInString } from 'codechum-app-utils';
import GLOBALS from 'codechum-app-globals';
import styles from './styles.module.scss';

import { Card, Text, Tags, Data, Badge } from '../../../elements';

import { textTypes } from '../../../elements/constants';

const ProblemCardSmall = ({
  score,
  problem: { name, id, tags },
  usedIn,
  viewProblem,
  onClick,
  disabled,
}) => (
  <Card
    className={styles.ProblemCardSmall}
    data-test="problemCardSmallContainer"
    id={`problemCardSmall_${id}`}
    onClick={!disabled ? onClick || viewProblem : () => {}}
  >
    <div className={styles.ProblemCardSmall_score}>
      <Data colorClass={GLOBALS.COLOR_CLASSES.BLUE['300']} label="Score">
        {score}
      </Data>
    </div>
    <div className={styles.ProblemCardSmall_title}>
      <Text type={textTypes.HEADING.XXS}>{name}</Text>
      <Tags
        className={styles.ProblemCardSmall_tags}
        id={tags.map((tag) => tag.id)}
        tags={tags.map((tag) => ({
          name: tag.name,
        }))}
      />
      {usedIn?.length > 0 && (
        <Badge
          className={styles.ProblemCardSmall_badge}
          colorName={GLOBALS.COLOR_NAMES.RED}
          text={getUsedInString(usedIn)}
        />
      )}
    </div>
  </Card>
);

ProblemCardSmall.defaultProps = {
  viewProblem: null,
  onClick: null,
  disabled: false,
};

ProblemCardSmall.propTypes = {
  score: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,

  problem: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string.isRequired,
    tags: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
      })
    ).isRequired,
  }).isRequired,

  disabled: PropTypes.bool,

  // the used in data of this problem with the user
  // in relation to the selected section
  usedIn: PropTypes.arrayOf(
    PropTypes.shape({
      problem_id: PropTypes.number,
      section: PropTypes.object,
      task: PropTypes.object,
    })
  ),

  viewProblem: PropTypes.func,

  // this is an optional override onclick function to this card
  onClick: PropTypes.func,
};

export default ProblemCardSmall;
