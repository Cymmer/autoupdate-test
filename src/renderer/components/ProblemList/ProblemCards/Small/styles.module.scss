@import '../../../../scss/variables';
@import '../../../../scss/mixins';
@import '../../../../scss/themes';

.ProblemCardSmall {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  padding: 16px 24px;

  &_title {
    display: flex;
    align-items: center;
  }

  &_tags {
    margin: 0 0 0 16px;

    @include breakpoint(sm) {
      display: none;
    }
  }

  &_badge {
    @include breakpoint(sm) {
      display: none;
    }
  }

  &_score {
    padding: 0 24px 0 0;
    margin: 0 24px 0 0;
    border-right: 1px solid;
    @include theme {
      border-color: t('borderColor');
    }
  }
}
