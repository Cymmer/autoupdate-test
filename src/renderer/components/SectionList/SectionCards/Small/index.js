import React from 'react';
import PropTypes from 'prop-types';
import { formatSectionSchedule, formatTeacherName } from 'codechum-app-utils';
import GLOBALS from 'codechum-app-globals';
import styles from './styles.module.scss';

import { CardLink, Text } from '../../../elements';
import { textTypes } from '../../../elements/constants';

import sectionIcons from '../../constants/sectionIcons';
import sectionColors from '../../constants/sectionColors';
import sectionListTypes from '../../constants/sectionListTypes';

import SectionImage from '../Image';

const SectionCardSmall = ({
  link,
  userType,
  section: {
    name,
    icon,
    color,
    schedules,
    course: { code: courseCode },
    teacher,
  },
}) => (
  <CardLink className={styles.SectionCardSmall} to={link}>
    <SectionImage
      className={styles.SectionCardSmall_image}
      color={
        color
          ? sectionColors[
              [...Object.keys(sectionColors)].find(
                (c) => sectionColors[c].value === color
              )
            ]
          : sectionColors.GREEN
      }
      icon={
        icon
          ? sectionIcons[
              [...Object.keys(sectionIcons)].find(
                (i) => sectionIcons[i].value === icon
              )
            ]
          : sectionIcons.CODY
      }
      size={sectionListTypes.SMALL}
    />
    <div className={styles.SectionCardSmall_info}>
      <Text
        className={styles.SectionCardLarge_title}
        colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['700']}
        type={textTypes.HEADING.XS}
      >
        {name}
      </Text>
      <Text
        colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['300']}
        type={textTypes.BODY.SM}
      >
        {userType === GLOBALS.USER_TYPES.STUDENT
          ? courseCode.concat(' · ', formatTeacherName(teacher))
          : courseCode.concat(' · ', formatSectionSchedule(schedules))}
      </Text>
    </div>
  </CardLink>
);

SectionCardSmall.propTypes = {
  // the link after this card is clicked
  link: PropTypes.string.isRequired,

  // the type of user this card is for
  userType: PropTypes.oneOf(Object.values(GLOBALS.USER_TYPES)).isRequired,

  // section object from List Sections API
  section: PropTypes.oneOfType([PropTypes.object]).isRequired,
};

export default SectionCardSmall;
