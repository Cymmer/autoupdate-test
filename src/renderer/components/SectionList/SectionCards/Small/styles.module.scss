@import '../../../../scss/variables';
@import '../../../../scss/mixins';

.SectionCardSmall {
  overflow: hidden;
  display: flex;

  &_image {
    width: 112px;
  }

  &_title {
    @include text-ellipsis;
  }

  &_info {
    padding: 16px 24px;
    width: calc(100% - 112px);
  }
}
