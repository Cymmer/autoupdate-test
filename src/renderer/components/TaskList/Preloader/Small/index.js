import React from 'react';
import styles from './styles.module.scss';

import { Card, Shine } from '../../../elements';

const PreloaderSmall = () => (
  <Card className={styles.PreloaderSmall}>
    <div className={styles.PreloaderSmall_image} />
    <div className={styles.PreloaderSmall_content}>
      <div className={styles.PreloaderSmall_shineHolder}>
        <Shine className={styles.PreloaderSmall_shine} />
        <Shine className={styles.PreloaderSmall_shine} />
      </div>
    </div>
  </Card>
);

export default PreloaderSmall;
