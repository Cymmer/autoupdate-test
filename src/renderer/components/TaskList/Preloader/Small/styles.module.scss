@import '../../../../scss/mixins';
@import '../../../../scss/animations';

.PreloaderSmall {
  display: flex;
  overflow: hidden;
  @include ease(0.4s);

  &_textContainer {
    display: flex;
    max-width: 80%;

    @include breakpoint(xs) {
      flex-direction: column;
      margin-bottom: 16px;
      max-width: 100%;
    }
  }

  &_image {
    width: 112px;
    @include image-preloader;
  }

  &_content {
    padding: 16px 24px;
    width: calc(100% - 112px);
  }

  &_info {
    display: flex;
    width: 100%;
    @include ease(0.2s);
  }

  &_shineHolder {
    width: 100%;
    @include flex-center;
    flex-direction: column;
  }

  &_shine {
    margin: 7px 0;
  }
}
