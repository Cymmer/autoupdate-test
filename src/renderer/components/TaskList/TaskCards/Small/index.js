import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import cn from 'classnames';
import moment from 'moment';
import {
  formatTaskSchedule,
  determinePercentageColorName,
} from 'codechum-app-utils';
import GLOBALS from 'codechum-app-globals';
import styles from './styles.module.scss';

import {
  Card,
  IconButton,
  Text,
  Data,
  Tooltip,
  IconLink,
} from '../../../elements';

import {
  textTypes,
  iconButtonTypes,
  tooltipPlacement,
} from '../../../elements/constants';

const TaskCardSmall = ({
  id,
  taskWithDetails,
  userType,
  link,
  viewLeaderboard,
  isExamOngoing,
}) => {
  const {
    name,
    is_disabled: isDisabled,
    start_time: startTime,
    end_time: endTime,
    is_leaderboard_hidden: isLeaderboardHidden,
    section: { name: sectionName } = {},
  } = taskWithDetails;

  return (
    <Card className={styles.TaskCardSmall} id={id}>
      <Link className={styles.TaskCardSmall_link} to={link}>
        {userType === GLOBALS.USER_TYPES.STUDENT && (
          <StudentSideData taskWithDetails={taskWithDetails} />
        )}
        <div
          className={cn({
            [styles.TaskCardSmall_content___student]:
              userType === GLOBALS.USER_TYPES.STUDENT,
            [styles.TaskCardSmall_content___teacher]:
              userType === GLOBALS.USER_TYPES.TEACHER,
          })}
        >
          <div>
            <Text
              className={styles.TaskCardSmall_title}
              colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['700']}
              type={textTypes.HEADING.XXS}
            >
              {name}
            </Text>
            <Text
              colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
              type={textTypes.BODY.SM}
            >
              {formatTaskSchedule(moment(startTime), moment(endTime))}
              {userType === GLOBALS.USER_TYPES.TEACHER && ` Â· ${sectionName}`}
            </Text>
          </div>
        </div>
      </Link>
      {userType === GLOBALS.USER_TYPES.STUDENT && (
        <div className={styles.TaskCardSmall_buttonContainer}>
          <Tooltip content="Review" placement={tooltipPlacement.BOTTOM}>
            <IconLink
              className={styles.TaskCardSmall_buttonContainer_button}
              disabled={!isDisabled || !isExamOngoing}
              icon="replay"
              to={
                !isDisabled && !isExamOngoing
                  ? `/student/review/${taskWithDetails.id}/question/1`
                  : '#'
              }
            />
          </Tooltip>

          <Tooltip
            content="View Leaderboard"
            placement={tooltipPlacement.BOTTOM}
          >
            <IconButton
              className={styles.TaskCardSmall_buttonContainer_button}
              disabled={isLeaderboardHidden === true}
              icon="assessment"
              iconButtonTypes={iconButtonTypes.SOLID.SM}
              onClick={(e) => {
                e.stopPropagation();
                if (!isLeaderboardHidden) {
                  viewLeaderboard(taskWithDetails);
                }
              }}
            />
          </Tooltip>
        </div>
      )}
    </Card>
  );
};

/* eslint-disable react/prop-types */
const StudentSideData = ({
  taskWithDetails: {
    max_score: maxScore,
    details: { score },
  },
}) => (
  <div
    className={
      styles[
        `TaskCardSmall_accent___${determinePercentageColorName(
          parseInt(score),
          maxScore
        )}`
      ]
    }
  >
    <Data
      className={styles.TaskCardSmall_data}
      colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['0']}
      label="Score"
      labelColorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['0']}
    >
      {parseInt(score) || 0}/{maxScore}
    </Data>
  </div>
);

TaskCardSmall.defaultProps = {
  id: null,
  viewLeaderboard: () => {},
  isExamOngoing: false,
};

TaskCardSmall.propTypes = {
  id: PropTypes.string,
  // task object from List Tasks API
  // details from List Student Tasks API
  taskWithDetails: PropTypes.oneOfType([PropTypes.object]).isRequired,

  // the type of user this card is for
  userType: PropTypes.oneOf([
    GLOBALS.USER_TYPES.STUDENT,
    GLOBALS.USER_TYPES.TEACHER,
  ]).isRequired,

  // the link where the user will be redirected when this is clicked
  link: PropTypes.string.isRequired,

  // opens the leaderboard modal for this task
  viewLeaderboard: PropTypes.func,

  // true if there's currently an exam ongoing
  isExamOngoing: PropTypes.bool,
};

export default TaskCardSmall;
