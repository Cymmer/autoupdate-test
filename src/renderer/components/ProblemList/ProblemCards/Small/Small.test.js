import React from 'react';
import { shallow } from 'enzyme';
import { findByTestAttr } from 'codechum-app-utils';
import Small from './index';

const DEFAULT_PROBLEM = {
  name: 'Dummy Problem',
  tags: [
    {
      id: 1,
      name: 'DummyTag1',
    },
    {
      id: 2,
      name: 'DummyTag2',
    },
  ],
};

describe('RENDERING - ProblemList > ProblemCards > Small', () => {
  it('should render with default props', () => {
    const component = shallow(<Small problem={DEFAULT_PROBLEM} score={10} />);
    expect(component).toMatchSnapshot();
  });

  it('should render the used in badge if the problem has already been used in other activities', () => {
    const component = shallow(
      <Small
        problem={DEFAULT_PROBLEM}
        score={10}
        usedIn={[
          // since the shape of the used in objects are not necessary in the
          // context of this card, we just add empty objects
          {
            problem_id: 1,
            section: {},
            task: {},
          },
        ]}
      />
    );
    expect(component).toMatchSnapshot();
  });
});

describe('EVENTS - ProblemList > ProblemCards > Small', () => {
  it('should be able to click the problem if there is an onclick handler and is not disabled', () => {
    const clickFn = jest.fn();
    const component = shallow(
      <Small
        disabled={false}
        problem={DEFAULT_PROBLEM}
        score={10}
        onClick={clickFn}
      />
    );

    findByTestAttr(component, 'problemCardSmallContainer').simulate('click');

    expect(clickFn).toHaveBeenCalled();
  });

  it('should be able to view the problem if there is an view problem handler handler and is not disabled', () => {
    const clickFn = jest.fn();
    const component = shallow(
      <Small
        disabled={false}
        problem={DEFAULT_PROBLEM}
        score={10}
        viewProblem={clickFn}
      />
    );

    findByTestAttr(component, 'problemCardSmallContainer').simulate('click');

    expect(clickFn).toHaveBeenCalled();
  });

  it('should not be able to view or click the problem if disabled', () => {
    const viewFn = jest.fn();
    const clickFn = jest.fn();
    const component = shallow(
      <Small
        disabled
        problem={DEFAULT_PROBLEM}
        score={10}
        viewProblem={viewFn}
        onClick={clickFn}
      />
    );

    findByTestAttr(component, 'problemCardSmallContainer').simulate('click');

    expect(viewFn).not.toHaveBeenCalled();
    expect(clickFn).not.toHaveBeenCalled();
  });
});
