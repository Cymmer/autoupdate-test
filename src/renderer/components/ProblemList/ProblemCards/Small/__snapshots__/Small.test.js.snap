// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - ProblemList > ProblemCards > Small should render the used in badge if the problem has already been used in other activities 1`] = `
<Card
  className="ProblemCardSmall"
  data-test="problemCardSmallContainer"
  id="problemCardSmall_undefined"
  onClick={null}
  onMouseEnter={null}
  onMouseLeave={null}
>
  <div
    className="ProblemCardSmall_score"
  >
    <Data
      className={null}
      colorClass="b___300"
      id={null}
      label="Score"
      labelColorClass="n___400"
      size="md"
    >
      10
    </Data>
  </div>
  <div
    className="ProblemCardSmall_title"
  >
    <Text
      className={null}
      colorClass="n___700"
      element={null}
      id={null}
      type="heading___xxs"
    >
      Dummy Problem
    </Text>
    <Tags
      className="ProblemCardSmall_tags"
      id={
        Array [
          1,
          2,
        ]
      }
      tags={
        Array [
          Object {
            "name": "DummyTag1",
          },
          Object {
            "name": "DummyTag2",
          },
        ]
      }
    />
    <Badge
      className="ProblemCardSmall_badge"
      colorName="red"
      icon={null}
      id={null}
      text="Used in 1 activity"
    />
  </div>
</Card>
`;

exports[`RENDERING - ProblemList > ProblemCards > Small should render with default props 1`] = `
<Card
  className="ProblemCardSmall"
  data-test="problemCardSmallContainer"
  id="problemCardSmall_undefined"
  onClick={null}
  onMouseEnter={null}
  onMouseLeave={null}
>
  <div
    className="ProblemCardSmall_score"
  >
    <Data
      className={null}
      colorClass="b___300"
      id={null}
      label="Score"
      labelColorClass="n___400"
      size="md"
    >
      10
    </Data>
  </div>
  <div
    className="ProblemCardSmall_title"
  >
    <Text
      className={null}
      colorClass="n___700"
      element={null}
      id={null}
      type="heading___xxs"
    >
      Dummy Problem
    </Text>
    <Tags
      className="ProblemCardSmall_tags"
      id={
        Array [
          1,
          2,
        ]
      }
      tags={
        Array [
          Object {
            "name": "DummyTag1",
          },
          Object {
            "name": "DummyTag2",
          },
        ]
      }
    />
  </div>
</Card>
`;
